tensorboard/data/server/reservoir.rs

```diff
@@ -231,6 +233,53 @@ impl<T, C: ReservoirControl> StageReservoir<T, C> {
         &self.staged_items[..]
     }
 
+    /// Checks whether a preemption has been staged.
+    ///
+    /// This is initially `false`; it changes to `true` whenever a record is preempted, and back to
+    /// `false` whenever the reservoir is committed.
+    pub fn staged_preemption(&self) -> bool {
+        self.staged_preemption
+    }
+
+    /// Preempts any records whose step does not precede the given step.
+    fn preempt(&mut self, step: Step) {
+        let staged_nonpreempted = self
+            .staged_items
+            .iter()
+            .enumerate()
+            .rev()
+            .find_map(|(i, (s, _))| if *s < step { Some(i + 1) } else { None })
+            .unwrap_or(0);
+        let staged_preempted = self.staged_items.len() - staged_nonpreempted;
+        if staged_preempted > 0 {
+            self.staged_items.truncate(staged_nonpreempted);
+            self.staged_preemption = true;
+        }
+        let mut preempted = staged_preempted;
+        if self.staged_items.is_empty() {
+            // Committed steps may have been preempted as well.
+            let committed_nonpreempted = self
+                .committed_steps
+                .iter()
+                .enumerate()
+                .rev()
+                .find_map(|(i, s)| if *s < step { Some(i + 1) } else { None })
+                .unwrap_or(0);
+            let committed_preempted = self.committed_steps.len() - committed_nonpreempted;
+            if committed_preempted > 0 {
+                self.committed_steps.truncate(committed_nonpreempted);
+                self.staged_preemption = true;
+                preempted += committed_preempted;
+            }
+        }
+        if preempted == 0 {
+            return; // No need to adjust `seen`.
+        }
+        let old_stored = self.committed_steps.len() + self.staged_items.len() + preempted;
+        let fac_preempted = (preempted as f64) / (old_stored as f64);
+        self.seen = (self.seen as f64 * (1.0 - fac_preempted)).ceil() as usize;
```

Comment:
@nfelt: When preempting back to the original step or before, I think
that it actually doesn’t have any rounding error as written.

Since `preempt` is called first in `offer`, before even incrementing
`seen`, when you preempt to step 0 we get `fac_preempted == 1.0` exactly
(since `preempted == old_stored`), and thus set `self.seen = 0`.

This test passes:

```rust
    #[test]
    fn test_preempt_long() {
        let mut rsv = StageReservoir::new(10);
        let mut head = Basin::new();
        for i in 0..100_000 {
            rsv.offer(Step(i), ());
        }
        rsv.commit(&mut head);
        // head.as_slice()[0] happens to be step 25630 here
        assert_eq!(rsv.seen, 100_000);
        rsv.offer(Step(0), ());
        assert_eq!(rsv.seen, 1); // exactly correct: just the newest record
        for i in 1..9 {
            rsv.offer(Step(i), ());
        }
        rsv.commit(&mut head);
        assert_eq!(
            head.as_slice()
                .iter()
                .map(|&(Step(s), _)| s)
                .collect::<Vec<i64>>(),
            (0..9).collect::<Vec<i64>>()
        );
        assert_eq!(rsv.seen, 9); // exactly correct: just 0 and 1..9
    }
```

Supposing that we include a test something like this, does it seem
reasonable to you to keep the code as is, without special casing for
preempting back to start?